lib/Models/CatalogSearchProvider.ts

```diff
@@ -0,0 +1,156 @@
+import { observable, runInAction } from "mobx";
+import SearchProvider from "terriajs/lib/Models/SearchProvider";
+import SearchResult from "terriajs/lib/Models/SearchResult";
+import Terria from "terriajs/lib/Models/Terria";
+import SearchProviderResults from "terriajs/lib/Models/SearchProviderResults";
+import GroupMixin from "../ModelMixins/GroupMixin";
+import ReferenceMixin from "../ModelMixins/ReferenceMixin";
+
+interface CatalogSearchProviderOptions {
+  terria: Terria;
+}
+
+type UniqueIdString = string;
+type ResultMap = Map<UniqueIdString, boolean>;
+export function loadAndSearchCatalogRecursively(
+  terria: Terria,
+  searchTextLowercase: string,
+  searchResults: SearchProviderResults,
+  resultMap: ResultMap,
+  iteration: number = 0
+): Promise<Terria> {
+  // checkTerriaAgainstResults(terria, searchResults)
+  // don't go further than 10 deep, but also if we have references that never
+  // resolve to a target, might overflow
+  if (iteration > 10) {
+    return Promise.resolve(terria);
+  }
+  // add some public interface for terria's `models`?
+  const referencesAndGroupsToLoad: any[] = Array.from(
+    (<any>terria).models.values()
+  ).filter((model: any) => {
+    if (resultMap.get(model.uniqueId) === undefined) {
+      const modelToSave = model.target || model;
+      // Use a flattened string of definition data later,
+      // without only checking name/id/descriptions?
+      // saveModelToJson(modelToSave, {
+      //   includeStrata: [CommonStrata.definition]
+      // });
+      const matchesString =
+        `${modelToSave.name} ${modelToSave.uniqueId} ${modelToSave.description}`
+          .toLowerCase()
+          .indexOf(searchTextLowercase) !== -1;
+      resultMap.set(model.uniqueId, matchesString);
+      if (matchesString) {
+        runInAction(() => {
+          searchResults.results.push(
+            new SearchResult({
+              name: name,
+              catalogItem: model
+            })
+          );
+        });
+      }
+    }
+
+    if (ReferenceMixin.is(model) || GroupMixin.isMixedInto(model)) {
+      return true;
+    }
+    // Could also check for loadMembers() here, but will be even slower
+    // (relies on external non-magda services to be performant)
+
+    return false;
+  });
+  if (referencesAndGroupsToLoad.length === 0) {
+    return Promise.resolve(terria);
+  }
+  return new Promise(resolve => {
+    Promise.all(
+      referencesAndGroupsToLoad.map(model => {
+        if (ReferenceMixin.is(model)) {
+          return model.loadReference();
+        } else if (GroupMixin.isMixedInto(model)) {
```

Comment:
Now I remember the reason why I didn't check groupmixin in DT, it meant search would get stuck on waiting for remote services outside of magda (as we can reliably resolve magda references quickly, but there's no guarantees for say a WMS group). Nevertheless we can give it a try and see how it goes with various catalogs